File: Java/484. Find Permutation.java
class Solution {
    public int[] findPermutation(String s) {
        int n = s.length()+1;
        Stack<Integer> stack = new Stack<>();
        int[] ans = new int[n];
        int j = 0;
        
        for (int i=1; i<n; i++) {
            if (s.charAt(i-1) == 'I') {
                stack.push(i);
                while (!stack.isEmpty()) {
                    ans[j++] = stack.pop();
                }
            } else {
                stack.push(i);
            }
        }
        
        stack.push(n);
        while (!stack.isEmpty()) {
            ans[j++] = stack.pop();
        } 
        
        return ans;
    }
}

File: Python3/484. Find Permutation.py
class Solution:
    def findPermutation(self, s: str) -> List[int]:
        n = len(s)+1
        stack = deque()
        ans = []
        
        for i in range(1, n):
            if s[i-1] == 'I':
                stack.append(i)
                while stack:
                    ans.append(stack.pop())
            else:
                stack.append(i)
            
        stack.append(n)
        while stack:
            ans.append(stack.pop())
            
        return ans